README.md
# Activity-Logger.***py***

*Basic Activity Logging Script*

---

## Briefing

_**Actlogger** is a command-line tool that can be used to establish records of your sequential day-to-day works in a log. This tool integrates well with projects like 50 and 100 days of code. Users can keep track of their work at the end of the day into this log to keep track of everything they learn over the course._

---

## Installation V0.0.4

step-1: ``requirements``

```bash
$ pip install -r requirements.txt
```

step-2: ``clone repo``

```bash
$ git clone https://github.com/r0han99/activity-logger-py.git
```

step-3: ``setup.sh``

```bash
$ cd into-repo #representation
$ chmod +x setup.sh
$ ./setup.sh
```

~ _Execlusive only to MacOS, Linux based OS_

---

## Usage

- **$** ``Establish``

  ```bash
  $ actlogger --init
  ```
  <img width="616" alt="image-20220214112615681" src="https://user-images.githubusercontent.com/45916202/153812852-1d1a5786-5369-4a62-8f2d-9202fa015c64.png">


* **$** ``Functionalities``

  ```bash
  $ actlogger --tasks constructors-classes-objects
  ```
  <img width="528" alt="image-20220214112959474" src="https://user-images.githubusercontent.com/45916202/153813027-ec9fe36e-186f-44f8-916a-d55906cb7f83.png">


   ```
  $ actlogger --tasks constructors-classes-objects --info y # for logging additional content
   ```

  <img width="1225" alt="image-20220214113454224" src="https://user-images.githubusercontent.com/45916202/153813139-9eb26b44-4333-40c3-91c3-74e0a82b3f7b.png">


* $ ``backup``

  ```bash
  $ actlogger --backup partial 
  ```

* **$** ``summary``

  ```bash
  $ actlogger --summary
  ```

  <img width="329" alt="image-20220214121803942" src="https://user-images.githubusercontent.com/45916202/153814416-397e849b-6f45-4900-a77a-bb60a31876cc.png">

 

---

#### Prospective Features

* ``--restore`` : Functionality to take the latest backedup Archive and Restore the logging setup, if the assets are deleted accidentally.
* ``--amend`` : Functionality to edit a particular ;og line.
* ``--visualise`` | ``--stats`` : Functionality to visualise statistics of logs. 
  * Topic or Concepts Mentioned more number of times.
  * Average Time Taken?
* ``--synch`` : Functionality to synch inconsistent records.
* ``--trackers`` : Functionality to keep track of all the active loggers present at the moment.(Multiple directories may have a tracker each)
* ``--streak-status``: Functionality to display all the times a consistent streak is broken, and what was the time-off. 
* A pip package? -- *Only time will tell.*

---



changelog.md
commit 4950385a12ba330fc91619d8b996fc116cf05363
Author: r0han99 <r409067@gmail.com>
Date:   Fri Feb 25 00:21:27 2022 +0530

    V0.0.4
    * Miniature Changes
    * Fixed Typos
    * Added Documentation at the beginning of every subroutine.
    * Added Changelog to the Repo

commit 192ce9b1a6dfa48782334a721407c46ee348ce11
Author: r0han99 <r409067@gmail.com>
Date:   Sun Feb 20 23:31:42 2022 +0530

    updated --version-sequence

commit aa120631571f034a628d62842815ecadb91fbeae
Author: r0han99 <r409067@gmail.com>
Date:   Sun Feb 20 23:25:14 2022 +0530

    updated shabang version

commit 49165f6dee373f0d136aa024d0487da4896d45c0
Author: Rohan Sai <45916202+r0han99@users.noreply.github.com>
Date:   Tue Feb 15 11:13:10 2022 +0530

    appended prospective features

commit 4cd36b9abc691787d2436fa6d2825dcf9a8c2d8a
Author: Rohan Sai <45916202+r0han99@users.noreply.github.com>
Date:   Mon Feb 14 12:23:34 2022 +0530

    Updated ReadMe

commit 025e6e01b1a33a830c671933eb171ac28d60f402
Author: r0han99 <r409067@gmail.com>
Date:   Mon Feb 14 12:22:37 2022 +0530

    fixed summary outputs

commit 2195d4ab4c817bcefdc5e9fb1ccdb113677e41d7
Author: Rohan Sai <45916202+r0han99@users.noreply.github.com>
Date:   Mon Feb 14 12:14:44 2022 +0530

    updated image-links

commit 478f481e11ea9200826c998f3c76cb203f5c36d2
Author: Rohan Sai <45916202+r0han99@users.noreply.github.com>
Date:   Mon Feb 14 12:06:13 2022 +0530

    Updated ReadMe

commit e284e7cdfd9f05eaaa077694ade03e41a7cf651a
Author: Rohan Sai <45916202+r0han99@users.noreply.github.com>
Date:   Mon Feb 14 12:05:26 2022 +0530

    updated ReadMe

commit 25054976680ac00390bba2981eb13deecfa909e9
Author: Rohan Sai <45916202+r0han99@users.noreply.github.com>
Date:   Mon Feb 14 12:04:12 2022 +0530

    updated ReadMe

commit ea2cee2f091263739f7b8148223754db0e95588d
Author: r0han99 <r409067@gmail.com>
Date:   Mon Feb 14 12:03:18 2022 +0530

    ++ usage-illustration.png

commit 3ba26b6b927527c8911ab2bf469d8b8f6ad67c89
Author: r0han99 <r409067@gmail.com>
Date:   Mon Feb 14 12:01:28 2022 +0530

    ReadMe Updated

commit a5fee70bdbfe0545059889335b02fcccf82b79d9
Author: r0han99 <r409067@gmail.com>
Date:   Sun Feb 13 22:59:57 2022 +0530

    V1.0 complete
    - completed setup.sh
    - summary functionality is completed
    - added sample archive structure to the test/ folder
    - pending ReadMe

commit f5e2b657be036443d5341e96bfc3d6ae256bd638
Author: r0han99 <r409067@gmail.com>
Date:   Sun Feb 13 22:20:28 2022 +0530

    V1.0 complete
    - setup.sh pending

commit 795f8bd5ae6a6b88e1efed0e770974ca1836190a
Author: r0han99 <r409067@gmail.com>
Date:   Sun Feb 13 19:26:00 2022 +0530

    V1.0 Amends
    - --init fucntionality to instantiate asset holders
    - multi-level hierachical checks inplace to makesure the control doesn't break when assets are missing
    - pending works
        - backup funcs
        - setup.sh
        - --summary flag fucntionality

commit ec679f2d68be1dafe5af4243d70de64c7eb025b2
Author: r0han99 <r409067@gmail.com>
Date:   Tue Feb 1 00:51:32 2022 +0530

    V1.0 - pending init & backup funcs

commit f32fffecf8d25a63614dc4d948099e43222b3784
Author: Rohan Sai <45916202+r0han99@users.noreply.github.com>
Date:   Tue Feb 1 00:36:56 2022 +0530

    Initial commit


setup.sh
#!/bin/bash


# confirmation
echo "This setup presumes that you have Python-3>= pre-installed on your PC along with Pip. Enter [y/n]: "
read control

# details 
BLUE='\033[0;36m'
NC='\033[0m' # No Color


if [[ $control == 'y' ]]
then
    echo "+ Instantiating Logger Functionality"
    echo "+ Establishing assets."

    # requirements
    pip install -r requirements.txt
    
    echo "+ Done."
    
    # storing the file path
    logger_script="./activity_logger.py"

    echo -e "+ Forging Command: ${BLUE}actlogger${NC}"
    touch actlogger
    cp $logger_script actlogger

    echo "+ converting script into executable"
    echo "+ writing permissions; enter password if prompted."
    # making it into an executable
    chmod +x actlogger

    # copied into execs
    sudo cp ./actlogger /usr/local/bin/
    echo -e "+ $ ${BLUE}actlogger${NC} is now forged as a commandline tool. Please try actlogger --help to initiate usage."



else
    echo "x Please install both Python and Pip package handler."
fi
    









activity_logger.py
#!/usr/bin/env python3.8

import shutil
import sys 
import os
import argparse
import numpy as np
import pandas as pd
from datetime import datetime
from zipfile import ZipFile
from tqdm import tqdm



# SCRIPT VERSION
VERSION = 0.0.3


# PARSER instantiation
parser = argparse.ArgumentParser()
parser.add_argument("--init", help='Initialises Archive directory structure', action='store_true')
parser.add_argument("-t", "--tasks", type=str, help='Enter all the key-topics formatted with hypens("-").')
parser.add_argument("-i", "--info", type=str, help="Input Y/N to enter additional details in the catalog.", choices=['Y','N','y','n'])
parser.add_argument("--summary", help='Displays summary of the most recent work.', action='store_true')
parser.add_argument("--backup", type=str, help='Backsup the Logdata with a dedicated timedelta.',choices=['partial','absolute'])
parser.add_argument('--version', action='version', help='Program version.', version=f'Activity Logger (actlogger) \033[96m(V{VERSION})\033[0m')
                    
args = parser.parse_args()

# print(args.init)
# print(args.info)
# print(args.tasks)
# print(args.backup)



abspath = os.getcwd()
path = os.path.join(abspath, '.archive')


# INIT
def init():
    
    global path

    # archive existence check
    if os.path.exists(path):
        print('- Archive exists')

        if os.path.exists(os.path.join(path,'logdata.csv')):
            print('- Log Database exists.')
    
        else:
            # create
            print("x Log Database doesn't exist.")
            header = "TID,DATE,DAY,TASK,DETAILS\n"
            with open(os.path.join(path,'logdata.csv'),'w') as f:
                f.write(header)
            print("- Log Database established.")
            
            

        if os.path.exists(os.path.join(path, 'tasksequence.log')):
            print('- Tasks Sequence LogFile exists.')

        else:
            # create
            print("x Task Sequence LogFile doesn't exist")
            with open(os.path.join(path, 'tasksequence.log'), 'w') as f:
                pass
            print("+ Task Sequence LogFile established.")

            
        # details folder
        if os.path.exists(os.path.join(path, 'details')):
            print('- Additional content repository exists.')
        else:
            print("x Additional content repository doesn't exist")
            os.mkdir(os.path.join(path, 'details'))
            print('+ Additional content repository established.')

        # backup folder
        if os.path.exists(os.path.join(path, 'backup')):
            print('- Backup repository exists.')
        else:
            print("x Backup repository doesn't exist")
            os.mkdir(os.path.join(path, 'backup'))
            print('+ Backup repository established.')

            
        if args.init == True:
            print('Setup Complete, use actlogger --help for functionality and assistance.')
        else:
            print('Setup Complete.')
        print('')
        


    else:
        
        
        # establish directory
        os.makedirs(path)
        print('- establishing the requirements.')
        # Log Database 
        header = "TID,DATE,DAY,TASK,DETAILS\n"
        with open(os.path.join(path,'logdata.csv'),'w') as f:
            f.write(header)
        print("+ Log Database established.")
    
        # LogFile
        with open(os.path.join(path, 'tasksequence.log'), 'w') as f:
            pass
        print("+ Task Sequence LogFile established.")


        # details folder
        os.mkdir(os.path.join(path, 'details'))
        print('+ Additional content repository established.')

         # backup folder
        os.mkdir(os.path.join(path, 'backup'))
        print('+ Backup repository established.')

        if args.init == True:
            print('Setup Complete, use activity_logger --help for functionality and assistance.')
        else:
            print('Setup Complete.')

            
        print('')


def fetch_records():

    global path 
    if os.path.exists(os.path.join(path,'logdata.csv')):

        # Global Day Counter : Initial 
        try:
            records = pd.read_csv(os.path.join(path,'logdata.csv').drop('Unnamed: 0',axis=1))
        except: 
            records = pd.read_csv(os.path.join(path,'logdata.csv'))
        try:
            day = records['DAY'][-1:].values[0]
        except:
            day = 0

        # secondary check 
        if 'Unnamed: 0' in records.columns:
            records = records.drop('Unnamed: 0',axis=1)
        

        return records, day
    else:

        print('x Missing required archive, executing --init to establish files.')
        init()
        print('*' * 25)
        records, day = fetch_records()
        return records, day






# create log-ID number for the datetime provided
def idgen(filepath):
    
    idnum = np.random.randint(1000,9999)

    # read-id sequence from csv
    numseq = pd.read_csv(filepath)['TID']

    while True:
        if idnum in numseq:
            idnum = np.random.randint(1000,9999)
            continue
        else:
            break
    return idnum

def package_details(tasks, info='N'):

    global path
    _, day = fetch_records()
    date = datetime.now().strftime("%Y-%m-%d-%H:%M:%S") 
    day = day+1
    idnum = idgen(os.path.join(path,'logdata.csv'))
    task = '; '.join(list(map(str.capitalize, args.tasks.split('-'))))
    template = "[{}] x {} -- ({}) -- day=[{}], files-created:{}".format(idnum,date,task,day,args.info)
    display_template =  "[{}] -- ({}) -- day=[{}]".format(idnum,task,day)

    packet = {"TID":idnum, "DATE":date, "DAY": day, "TASK":task, "DETAILS": info}

    return packet, template, display_template


def write_data(packet,records,template):

    global path
    
    # Write Data into DataFrame
    row = list(packet.values())
    # print(row)
    records.loc[len(records.index)] = row
    records.to_csv(os.path.join(path,'logdata.csv'))

    # Logfile
    with open(os.path.join(path,'tasksequence.log'),'a+') as f:
        f.write(template+'\n')

    
def backup_suggestion(day):

    if day%10 == 0:
        print('Partial backup recommended.')
    elif day%25 == 0:
        print('Full backup recommended')

    # Next Version
    # elif day%50 == 0:
    #     print('commencing Full Backup')

    else:
        pass



def silent_check():

    abspath = os.getcwd()
    path = os.path.join(abspath, '.archive')
    
    # archive existence check
    if os.path.exists(path):
        # print('- Archive Exists')

        if os.path.exists(os.path.join(path,'logdata.csv')):
            # print('- Log Database exists.')
            pass
    
        else:
            # create
            print("x \033[91m[FATAL]\033[0m Log Database doesn't exist.")
            header = "TID,DATE,DAY,TASK,DETAILS\n"
            with open(os.path.join(path,'logdata.csv'),'w') as f:
                f.write(header)
            print("- New Log Database established.")
            print("x \033[93m[WARNING]\033[0m Day Tracking Disrupted!. Day count will start from 1")

            '''
                Write a functionality or an addtional flag prompt to synchronise the day counter.
            
            '''
            print('-- Proceeding')
            print('')
            
            

        if os.path.exists(os.path.join(path, 'tasksequence.log')):
            # print('- Tasks Sequence LogFile exists.')
            pass

        else:
            # create
            print("x \033[91m[FATAL]\033[0m Task Sequence LogFile doesn't exist")
            with open(os.path.join(path, 'tasksequence.log'), 'w') as f:
                pass
            print("+ Task Sequence LogFile established.")
        
            print('-- Proceeding')
            print('')

        # details folder
        if os.path.exists(os.path.join(path, 'details')):
            # print('- Additional content repository exists.')
            pass
        else:
            print("x \033[91m[FATAL]\033[0m Additional content repository doesn't exist")
            os.mkdir(os.path.join(path, 'details'))
            print('+ Additional content repository established.')

            
            print('-- Proceeding')
            print('')

        # backup folder
        if os.path.exists(os.path.join(path, 'backup')):
            # print('- Additional content repository exists.')
            pass
        else:
            print("x \033[91m[FATAL]\033[0m Backup repository doesn't exist")
            os.mkdir(os.path.join(path, 'backup'))
            print('+ Backup repository established.')

            
            print('-- Proceeding')
            print('')

        '''
        Display a Note to inform the user that if at all after the FATAL prompt the data logged is not the first, then in that case
        user should be prompted to use --restore (V.2)
        
        '''
        
        
## ARGUMENT PROCESSING

funcs = {}
try:
    funcs['init'] = args.init
    funcs['tasks'] = args.tasks
    funcs['info'] = args.info
    funcs['summary'] = args.summary
    funcs['backup'] = args.backup
    funcs['version'] = args.version
except:
    pass

# print(funcs)

if not funcs['tasks'] == None:
    
    # silent check for all the requirced file's existence
    silent_check()

    # fetching records
    records, day = fetch_records()
    packet, template, display_template = package_details(args.tasks, args.info)
    DAY = packet['DAY']

    #backup prompt check
    backup_suggestion(DAY)

    # print logged information
    print('+',display_template, end='--')
    
    packet['DETAILS'] = 'n'
    write_data(packet,records,template)
    
    print('\033[96m logged \033[0m')

    if not funcs['info'] == None:

        if args.info in ['Y','y']:
            
            idnum = packet['TID']
            print(f'-- Additional Content [{idnum}]-- ')
            content = input('Enter content to be entered: ')
            print(f'File Created - {idnum}.txt')
            
            with open(os.path.join(path, f'details/{idnum}.txt'),'w') as f:
                f.write(content)
            
            print('+ Addtional content',end='')
            print('\033[96m logged \033[0m')
            

elif not funcs['backup'] == None:

    # check if the logs are absolutely empt
    # check if the logfile and the logdatabase have atleast 20 Logs else prompt to override the backy which doesn't require any backup procedureup procedure
    

    # to make sure everything's in place
    silent_check()

    records, _ = fetch_records()
    control = 'proceed'

    # initial checks
    if len(records) == 0:
        print('x No Records exists to be backed up.')
        control = 'halt'
    elif len(records) <= 20:
        print('~ Logs are limited, not in the optimal backup range. ')
        while True:
            flag = input('\nEnter `override` or `stop` to control : ')
            if flag == 'override':
                print('+ overriding backup, mode = ', end='')
                control = 'proceed'
                break
            elif flag == 'stop':
                print('stopping backup')
                control = 'halt'
                break
            else:
                print('x Invalid control choice!')
                
        
    if control == 'proceed':
    
        if args.backup == 'partial':
            # mode
            print('Partial')
            logdatapath = os.path.join(path, 'logdata.csv')
            tasksequencepath = os.path.join(path, 'tasksequence.log')

            
            # Generating Filename -- Time Delta
            dt = str(datetime.now().strftime('%Y-%m-%d %H:%M:%S'))
            dt = dt.replace('.','x')
            dt = '_'.join(dt.split(' '))


            # zipping
            tqdmprompts = [logdatapath,tasksequencepath]
            arcnames = ['logdata.csv','tasksequence.csv']
            cprname = dt+'.zip'
            with ZipFile(os.path.join(path,cprname), 'w') as zipper:
                for cprfile in tqdm(tqdmprompts, desc='compressing'):
                    zipper.write(cprfile, arcname=cprfile.split('/')[-1:][0])
                    
            # move the generated zip into the backup folder
            shutil.move(os.path.join(path,cprname),os.path.join(path,'backup/'))

        
        elif args.backup == 'absolute':
            print('Absolute Backup')
            print('- under development')

       


elif args.init:

    # check if the archive/ logdata.csv and tasksequence.log exists; if Yes then terminate else establish
    init()

elif args.summary:

    # fetch records
    records, day = fetch_records()
    
    if len(records) != 0:
        print('+ \033[96mSummary\033[0m')
        print(f'- Streak - Days: [{day}]')
        print(f"- Most Recent Work - [{records['TID'].values[-1:][0]}] : [{records['TASK'].values[-1:][0]}]")
        print(f"- Date Logged- [{records['DATE'].values[-1:][0]}]")

        '''
        Additional Info exists print that 
        '''


    else:
        print('x No Records to summarise.')

elif not args.info == None:
    print('x [invalid usage] --info is used in conjuction to --tasks. try --help for assistance.')

elif args.init == False and args.backup == None and args.info == None and args.tasks == None:
    print('x Try seeking assitance with --help.') 


    

        
    
